### PR TITLE
Protocol spec for uniform delegate shuffling - Closes #4922

### LINKED ISSUE #4922: Create protocol spec for shuffling of delegate
### Description
We need to create protocol spec for the introduction of LIP [Uniform ordering of delegates list](https://github.com/LiskHQ/lips/blob/master/proposals/lip-0003.md)

### Motivation
- To validate the implementation of LIP [Uniform ordering of delegates list](https://github.com/LiskHQ/lips/blob/master/proposals/lip-0003.md)
### Acceptance Criteria
- Create protocol-spec for shuffling of active 101 delegate

### Related issues
#4915

### PR DESCRIPTION
**Pull request by ishantiw**
### What was the problem?

This PR resolves #4922 

### How was it solved?

Scenario
- For a given list of delegates, provides a uniformly ordered delegate list

Created protocol specs for the implementation of LIP [Uniform ordering of delegates list](https://github.com/LiskHQ/lips/blob/master/proposals/lip-0003.md)

### How was it tested?

`node protocol-specs/generators/dpos_delegate_shuffling/index.js`

### CODE REVIEW

File: protocol-specs/generators/dpos_delegate_shuffling/index.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+const { hash } = require('@liskhq/lisk-cryptography');
+const BaseGenerator = require('../base_generator');
+const previousDelegateList = require('./delegate_address_list.json').list;
+
+const generateShuffledDelegateList = () => {
+	const previousRoundSeed = 'b9acc2f1fda3666bfb34107f1c6dccc4';
+	const delegateList = [...previousDelegateList];
+	for (const delegate of delegateList) {
+		const seedSource = previousRoundSeed + delegate.address;
```
Comment (shuse2) [inline protocol-specs/generators/dpos_delegate_shuffling/index.js]: I think this should be
```
const addressBuffer = Buffer.alloc(8);
addressBuffer.writeBigInt64BE(BitInt(delegate.address.slice(0, -1)));
const seedSource = Buffer.concat([Buffer.from(previousRoundSeed, 'hex'), addressBuffer]);
```

File: protocol-specs/generators/dpos_delegate_shuffling/index.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		delegate.roundHash = hash(seedSource, 'utf8').toString('hex');
```
Comment (shuse2) [inline protocol-specs/generators/dpos_delegate_shuffling/index.js]: then this becomes `delegate.roundHash = hash(seedSource)`

File: protocol-specs/generators/dpos_delegate_shuffling/index.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	}
+
+	delegateList.sort((delegate1, delegate2) => {
+		if (delegate1.roundHash !== delegate2.roundHash) {
+			return delegate1.roundHash.localeCompare(delegate2.roundHash);
```
Comment (shuse2) [inline protocol-specs/generators/dpos_delegate_shuffling/index.js]: if `roundHash` is buffer, then you can simple do 
```
const diff = delegate1.roundHash.compare(delegate2.roundHash);
if (diff !== 0) {
return diff;
}
return delegate1.address.localeCompare(delegate2.address);
```

File: protocol-specs/generators/dpos_delegate_shuffling/index.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+		}
+		return delegate1.address.localeCompare(delegate2.address);
+	});
+
+	return {
+		input: {
+			previousRoundSeed,
+			delegateList: previousDelegateList,
```
Comment (shuse2) [inline protocol-specs/generators/dpos_delegate_shuffling/index.js]: as for the input, array of address should be fine

File: protocol-specs/generators/dpos_delegate_shuffling/index.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+		},
+		output: {
+			delegateList,
```
Comment (shuse2) [inline protocol-specs/generators/dpos_delegate_shuffling/index.js]: same as above

File: protocol-specs/generators/dpos_delegate_shuffling/index.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	const delegateList = [...previousDelegateList];
+	for (const delegate of delegateList) {
+		const addressBuffer = Buffer.alloc(8);
+		addressBuffer.writeBigUInt64BE(BigInt(delegate.address.slice(0, -1)));
+		const seedSource = Buffer.concat([
+			Buffer.from(previousRoundSeed1, 'hex'),
+			addressBuffer,
+		]);
+		/* eslint-disable dot-notation */
+		delegate['roundHash'] = hash(seedSource);
```
Comment (shuse2) [inline protocol-specs/generators/dpos_delegate_shuffling/index.js]: btw why not `dot` notation here?

Comment (ishantiw) [reply protocol-specs/generators/dpos_delegate_shuffling/index.js]: I changed it, I had TS in my mind before

File: protocol-specs/generators/dpos_delegate_shuffling/sample_generator.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+'use strict';
+
+const { getAddressFromPassphrase } = require('@liskhq/lisk-cryptography');
+const { Mnemonic } = require('@liskhq/lisk-passphrase');
+const fs = require('fs');
+
+const [amount = 101, filePath] = process.argv.slice(2);
```
Comment (ManuGowda) [inline protocol-specs/generators/dpos_delegate_shuffling/sample_generator.js]:
```suggestion
const [numberOfDelegates = 101, filePath] = process.argv.slice(2);
```

File: protocol-specs/generators/dpos_delegate_shuffling/sample_generator.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+		delegateList.push({
+			address,
+			roundHash: '',
+		});
+	}
+
+	return { list: delegateList };
+};
+
+const delegates = generateDelegates(amount);
```
Comment (ManuGowda) [inline protocol-specs/generators/dpos_delegate_shuffling/sample_generator.js]:
```suggestion
const delegates = generateDelegates(numberOfDelegates);
```

**Pull request merged**